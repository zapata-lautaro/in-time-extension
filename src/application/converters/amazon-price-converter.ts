import { Currency } from '../../models/currency.enum';
import {
  ORIGINAL_VALUE_ATTRIBUTE,
  PriceConverter,
  PriceSelector,
} from './price-converter';

const PRICE_VALUE_SELECTORS: PriceSelector[] = [
  {
    selector: '.a-price-whole',
    includeSymbol: false,
  },
  {
    selector: '.a-text-price span:last-child',
    includeSymbol: false,
  },
];
const PRICE_SYMBOL_SELECTORS = ['.a-price-symbol'];
const CENTS_SELECTORS = ['.a-price-fraction'];
const DECIMAL_SEPARATOR = '.';
const ARS_KEYWORD = 'ARS';

export class AmazonPriceConverter extends PriceConverter {
  constructor(document: Document) {
    super(
      document,
      PRICE_VALUE_SELECTORS,
      PRICE_SYMBOL_SELECTORS,
      CENTS_SELECTORS,
      Currency.USD,
      DECIMAL_SEPARATOR,
    );
  }

  protected override priceElementCurrencyGetter(element: Element): Currency {
    let isPriceInArs = element.parentElement
      .querySelector(PRICE_SYMBOL_SELECTORS[0])
      ?.textContent.toUpperCase()
      ?.includes(ARS_KEYWORD);

    isPriceInArs = isPriceInArs || element.textContent?.includes(ARS_KEYWORD);

    return isPriceInArs ? Currency.ARS : Currency.USD;
  }
}
