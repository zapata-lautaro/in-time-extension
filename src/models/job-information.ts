import { Currency } from './currency.enum';

export interface JobInformationProps {
  hoursPerMonth: number;
  hoursPerLaboralDay: number;
  salary: number;
  currency: Currency;
}

export class JobInformation implements JobInformationProps {
  constructor(
    public hoursPerMonth: number,
    public hoursPerLaboralDay: number,
    public salary: number,
    public currency: Currency,
  ) {}

  public static fromJson(json: JobInformationProps): JobInformation {
    return new JobInformation(
      json.hoursPerMonth,
      json.hoursPerLaboralDay,
      json.salary,
      json.currency,
    );
  }

  public anualSalary(): number {
    return this.salary * 12;
  }

  public monthlySalary(): number {
    return this.salary;
  }

  public weeklySalary(): number {
    return this.salary / 4;
  }

  public dailySalary(): number {
    return (this.salary / this.hoursPerMonth) * this.hoursPerLaboralDay;
  }

  public hourlySalary(): number {
    return this.salary / this.hoursPerMonth;
  }

  public perMinuteSalary(): number {
    return this.hourlySalary() / 60;
  }
}
