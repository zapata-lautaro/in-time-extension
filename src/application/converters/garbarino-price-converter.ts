import { Currency } from '../../models/currency.enum';
import { JobInformation } from '../../models/job-information';
import {
  PriceConverter,
  CONVERTED_TAG,
  ORIGINAL_VALUE_ATTRIBUTE,
  TIME_SYMBOL,
} from './price-converter.interface';

const PRICE_VALUE_SELECTORS = [
  '.price span:last-child',
  '.text-no-wrap.grey--text.font-2 span:last-child',
  '.discount-tag-price',
  'div[alternativediscountposition] .text-no-wrap.grey--text span:last-child',
];
const PRICE_SYMBOL_SELECTORS = [
  '.price span:nth-child(2)',
  '.text-no-wrap.grey--text.font-2 span:nth-child(2)',
  'div[alternativediscountposition] .text-no-wrap.grey--text span:nth-child(2)',
];

export class GarbarinoPriceConverter implements PriceConverter {
  constructor(private _document: Document) {}

  async convert(jobInformation: JobInformation): Promise<void> {
    if (jobInformation.hasMissingData()) {
      return;
    }

    await this.replacePrices(jobInformation);
    await this.replaceSymbols();
  }

  async revert(): Promise<void> {
    await this.revertConvertedElements();
  }

  private async revertConvertedElements() {
    const convertedElements = this._document.querySelectorAll(
      `[${ORIGINAL_VALUE_ATTRIBUTE}]`,
    );
    convertedElements.forEach((convertedElement) => {
      const originalValue = convertedElement.getAttribute(
        ORIGINAL_VALUE_ATTRIBUTE,
      );
      convertedElement.removeAttribute(ORIGINAL_VALUE_ATTRIBUTE);
      convertedElement.textContent = originalValue;
    });
  }

  private async replacePrices(jobInformation: JobInformation): Promise<void> {
    const filteredPrices = this._document.querySelectorAll(
      PRICE_VALUE_SELECTORS.join(','),
    );
    const prices = Array.from(filteredPrices).filter((element) => {
      return !element.querySelector(CONVERTED_TAG);
    });

    if (!prices) return;

    await Promise.all(
      Array.from(prices, async (priceElement) => {
        const originalValue = priceElement!.textContent;
        const regex = /(?<!,)\b\d+/g;
        const matches = originalValue.match(regex);
        const price = +matches.join('');
        const priceConvertion = jobInformation.getTimeConvertion(
          price,
          Currency.ARS,
        );

        return this.convertElement(
          priceElement,
          originalValue,
          priceConvertion.toShortStringRepresentation(),
        );
      }),
    );
  }

  private async replaceSymbols(): Promise<void> {
    const filteredCurrencySimbols = this._document.querySelectorAll(
      PRICE_SYMBOL_SELECTORS.join(','),
    );
    const currencySimbols = Array.from(filteredCurrencySimbols).filter(
      (element) => {
        return !element.querySelector(CONVERTED_TAG);
      },
    );
    await Promise.all(
      Array.from(currencySimbols, (symbolElement) => {
        const originalValue = symbolElement.textContent;
        this.convertElement(symbolElement, originalValue, TIME_SYMBOL);
      }),
    );
  }

  private async convertElement(
    element: Element,
    originalContent: string,
    newContent: string,
  ) {
    element.setAttribute(ORIGINAL_VALUE_ATTRIBUTE, originalContent);
    element.innerHTML = newContent + `<${CONVERTED_TAG}>`;
  }
}
