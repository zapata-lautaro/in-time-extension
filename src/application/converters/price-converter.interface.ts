import { JobInformation } from '../../models/job-information';

export interface PriceConverter {
  convert(jobInformation: JobInformation): Promise<void>;
  revert(): Promise<void>;
}
