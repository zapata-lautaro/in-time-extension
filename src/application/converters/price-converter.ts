import { Currency } from '../../models/currency.enum';
import { JobInformation } from '../../models/job-information';

export const TIME_SYMBOL = `⏱️`;
export const ORIGINAL_VALUE_ATTRIBUTE = 'data-original-value';
export const CONVERTED_TAG = 'converted';
export const HIDE_CLASS = 'time-convertion-hide';
export const DEFAULT_DECIMAL_SEPARATOR = ',';

export interface PriceSelector {
  selector: string;
  includeSymbol: boolean;
}

export abstract class PriceConverter {
  constructor(
    private _document: Document,
    private _priceValueSelectors: PriceSelector[],
    private _priceSymbolSelectors: string[],
    private _centsSelectors: string[],
    private _defaultCurrency: Currency,
    private _decimalSeparator: string = DEFAULT_DECIMAL_SEPARATOR,
  ) {}

  public async convert(jobInformation: JobInformation): Promise<void> {
    if (jobInformation.hasMissingData()) {
      return;
    }

    await this.replacePrices(jobInformation);
    await this.replaceSymbols();
    await this.removeCents();
  }

  async revert(): Promise<void> {
    await this.revertConvertedElements();
    await this.showCents();
  }

  private async revertConvertedElements() {
    const convertedElements = this._document.querySelectorAll(
      `[${ORIGINAL_VALUE_ATTRIBUTE}]`,
    );
    convertedElements.forEach((convertedElement) => {
      const originalValue = convertedElement.getAttribute(
        ORIGINAL_VALUE_ATTRIBUTE,
      );
      convertedElement.removeAttribute(ORIGINAL_VALUE_ATTRIBUTE);
      convertedElement.textContent = originalValue;
    });
  }

  private async replacePrices(jobInformation: JobInformation): Promise<void> {
    this._priceValueSelectors.forEach(async (priceSelector) => {
      const filteredPrices = this._document.querySelectorAll(
        priceSelector.selector,
      );
      const prices = Array.from(filteredPrices).filter((element) => {
        return !element.querySelector(CONVERTED_TAG);
      });

      await Promise.all(
        Array.from(prices, async (priceElement) => {
          const originalValue = priceElement!.textContent;
          const holeNumber = originalValue.includes(this._decimalSeparator)
            ? originalValue.split(this._decimalSeparator)[0]
            : originalValue;
          const matches = holeNumber.match(/\d+/g);
          const price = +matches?.join('') ?? 0;
          const priceConvertion = jobInformation.getTimeConvertion(
            price,
            this.priceElementCurrencyGetter(priceElement),
          );

          return this.convertElement(
            priceElement,
            originalValue,
            (priceSelector.includeSymbol ? `${TIME_SYMBOL} ` : '') +
              priceConvertion.toShortStringRepresentation(),
            `($${price}) ${priceConvertion.toFullStringRepresentation()}`,
          );
        }),
      );
    });
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  protected priceElementCurrencyGetter(element: Element): Currency {
    return this._defaultCurrency;
  }

  private async replaceSymbols(): Promise<void> {
    if (this._priceSymbolSelectors.length == 0) return;

    const filteredCurrencySimbols = this._document.querySelectorAll(
      this._priceSymbolSelectors.join(','),
    );
    const currencySimbols = Array.from(filteredCurrencySimbols).filter(
      (element) => {
        return !element.querySelector(CONVERTED_TAG);
      },
    );
    await Promise.all(
      Array.from(currencySimbols, (symbolElement) => {
        const originalValue = symbolElement.textContent;
        this.convertElement(symbolElement, originalValue, TIME_SYMBOL);
      }),
    );
  }

  private async removeCents() {
    if (this._centsSelectors.length == 0) return;

    const cents = this._document.querySelectorAll(
      this._centsSelectors
        .map((selector) => `${selector}:not(.${HIDE_CLASS})`)
        .join(','),
    );
    cents.forEach((cent) => {
      cent.classList.add(HIDE_CLASS);
    });
  }

  private async showCents() {
    if (this._centsSelectors.length == 0) return;

    const cents = this._document.querySelectorAll(
      this._centsSelectors.join(','),
    );
    cents.forEach((cent) => {
      cent.classList.remove(HIDE_CLASS);
    });
  }

  private async convertElement(
    element: Element,
    originalContent: string,
    newContent: string,
    tooltipContent?: string,
  ) {
    element.setAttribute(ORIGINAL_VALUE_ATTRIBUTE, originalContent);
    element.innerHTML = newContent + `<${CONVERTED_TAG}></${CONVERTED_TAG}>`;

    if (!tooltipContent) return;

    element.addEventListener('mouseover', function () {
      const parentRect = this.getBoundingClientRect();
      const tooltip = document.createElement('tool-tip');
      tooltip.innerHTML = tooltipContent;
      document.body.insertAdjacentElement('afterbegin', tooltip);
      tooltip.style.top = `${parentRect.bottom}px`;
      tooltip.style.left = `${parentRect.left}px`;
      tooltip.style.display = 'block';
    });

    element.addEventListener('mouseout', function () {
      const tooltip = document.querySelector('tool-tip');
      tooltip.remove();
    });
  }
}
