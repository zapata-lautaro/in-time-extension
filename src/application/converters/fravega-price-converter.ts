import { Currency } from '../../models/currency.enum';
import { PriceConverter, PriceSelector } from './price-converter';

const PRICE_VALUE_SELECTORS: PriceSelector[] = [
  {
    selector:
      'div[data-test-id="product-price"] span:not([data-test-id="discount-tag"])',
    includeSymbol: true,
  },
  {
    selector: 'div[data-test-id="price-wrapper"] span:not(:nth-child(3))',
    includeSymbol: true,
  },
];

export class FravegaPriceConverter extends PriceConverter {
  constructor(document: Document) {
    super(document, PRICE_VALUE_SELECTORS, [], [], Currency.ARS);
  }
}
