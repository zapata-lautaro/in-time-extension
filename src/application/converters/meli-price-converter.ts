import { Currency } from '../../models/currency.enum';
import { PriceConverter, PriceSelector } from './price-converter';

const PRICE_VALUE_SELECTORS: PriceSelector[] = [
  { selector: '.andes-money-amount__fraction', includeSymbol: false },
  { selector: '.dynamic-carousel__price span', includeSymbol: false },
];
const PRICE_SYMBOL_SELECTORS = ['.andes-money-amount__currency-symbol'];
const CENTS_SELECTORS = ['.andes-money-amount__cents'];
const USD_KEYWORD = 'd√≥lares';

export class MeliPriceConverter extends PriceConverter {
  constructor(document: Document) {
    super(
      document,
      PRICE_VALUE_SELECTORS,
      PRICE_SYMBOL_SELECTORS,
      CENTS_SELECTORS,
      Currency.ARS,
    );
  }

  protected override priceElementCurrencyGetter(element: Element): Currency {
    const isPriceInDolars = element.parentElement
      ?.getAttribute('aria-label')
      ?.includes(USD_KEYWORD);

    return isPriceInDolars ? Currency.USD : Currency.ARS;
  }
}
