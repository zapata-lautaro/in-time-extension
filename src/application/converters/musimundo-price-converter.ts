import { Currency } from '../../models/currency.enum';
import { PriceConverter, PriceSelector } from './price-converter';

const PRICE_VALUE_SELECTORS: PriceSelector[] = [
  {
    selector: '.mus-pro-price-save',
    includeSymbol: true,
  },
  {
    selector: '.mus-pro-price-price',
    includeSymbol: true,
  },
  {
    selector: '.mus-pro-price-number span',
    includeSymbol: true,
  },
  {
    selector: '.mus-pro-quotes-price',
    includeSymbol: false,
  },
];
const PRICE_SYMBOL_SELECTORS = ['.mus-pro-quotes-currency'];
const CENTS_SELECTORS = ['.mus-pro-quotes-decimals'];

export class MusimundoPriceConverter extends PriceConverter {
  constructor(document: Document) {
    super(
      document,
      PRICE_VALUE_SELECTORS,
      PRICE_SYMBOL_SELECTORS,
      CENTS_SELECTORS,
      Currency.ARS,
    );
  }
}
