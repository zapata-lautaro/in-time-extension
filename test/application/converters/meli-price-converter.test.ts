/**
 * @jest-environment jsdom
 */

import { MeliPriceConverter } from '../../../src/application/converters/meli-price-converter';
import { Currency } from '../../../src/models/currency.enum';
import { JobInformation } from '../../../src/models/job-information';

const meliNoPrices = `
    <span class="andes-money-amount" aria-label="1000000 pesos"></span>
`;

const originalMeliPesosHtml = `
    <span class="andes-money-amount" aria-label="1000000 pesos">
        <span class="andes-money-amount__currency-symbol">$</span>
        <span class="andes-money-amount__fraction">1.000.000</span>
        <span class="andes-money-amount__cents">44</span>
    </span>
`;

const originalMeliDolaresHtml = `
    <span class="andes-money-amount" aria-label="500 dólares">
        <span class="andes-money-amount__currency-symbol">$</span>
        <span class="andes-money-amount__fraction">500</span>
        <span class="andes-money-amount__cents">44</span>
    </span>
`;

const convertedMeliHtml = `
    <span class="andes-money-amount" aria-label="1000000 pesos">
        <span class="andes-money-amount__currency-symbol time-convertion">⏱️</span><span class="andes-money-amount__currency-symbol original">$</span>
        <span class="andes-money-amount__fraction time-convertion">10 Meses</span><span class="andes-money-amount__fraction original">1.000.000</span>
        <span class="andes-money-amount__cents original">44</span>
    </span>
`;

describe('convert', () => {
  const jobInformation = new JobInformation(5, 8, Currency.ARS, 100000, 1000);

  it('should replace prices and currency simbols', () => {
    document.body.innerHTML = originalMeliPesosHtml;

    const meliPriceConverter = new MeliPriceConverter(document, jobInformation);
    meliPriceConverter.convert();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('should replace prices in usd and currency simbols', () => {
    document.body.innerHTML = originalMeliDolaresHtml;

    const meliPriceConverter = new MeliPriceConverter(document, jobInformation);
    meliPriceConverter.convert();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('should not modify nothing if ther arent prices', () => {
    document.body.innerHTML = meliNoPrices;

    const meliPriceConverter = new MeliPriceConverter(document, jobInformation);
    meliPriceConverter.convert();

    expect(document.body.innerHTML).toBe(meliNoPrices);
  });
});

describe('revert', () => {
  const jobInformation = new JobInformation(5, 8, Currency.ARS, 100000, 1000);

  it('should remove converted elements and remove classes from original ones', () => {
    document.body.innerHTML = convertedMeliHtml;

    const meliPriceConverter = new MeliPriceConverter(document, jobInformation);
    meliPriceConverter.revert();

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
