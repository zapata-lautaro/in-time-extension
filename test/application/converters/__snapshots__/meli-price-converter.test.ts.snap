// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert should replace prices and currency simbols 1`] = `
"
    <span class="andes-money-amount" aria-label="1000000 pesos">
        <span class="andes-money-amount__currency-symbol" data-original-value="$">⏱️<converted></converted></span>
        <span class="andes-money-amount__fraction" data-original-value="1.000.000">10 Meses<converted></converted></span>
        <span class="andes-money-amount__cents time-convertion-hide">44</span>
    </span>
"
`;

exports[`convert should replace prices in usd and currency simbols 1`] = `
"
    <span class="andes-money-amount" aria-label="500 dólares">
        <span class="andes-money-amount__currency-symbol" data-original-value="U$D">⏱️<converted></converted></span>
        <span class="andes-money-amount__fraction" data-original-value="500">2 Semanas<converted></converted></span>
        <span class="andes-money-amount__cents time-convertion-hide">44</span>
    </span>
"
`;

exports[`revert should leave document as orignial without attributes or classes 1`] = `
"
    <span class="andes-money-amount" aria-label="500 dólares">
        <span class="andes-money-amount__currency-symbol">U$D</span>
        <span class="andes-money-amount__fraction">500</span>
        <span class="andes-money-amount__cents">44</span>
    </span>
"
`;
